source/common/secret/sds_api.cc

```diff
@@ -0,0 +1,89 @@
+#include "common/secret/sds_api.h"
+
+#include <unordered_map>
+
+#include "envoy/api/v2/auth/cert.pb.validate.h"
+
+#include "common/config/resources.h"
+#include "common/config/subscription_factory.h"
+#include "common/protobuf/utility.h"
+#include "common/ssl/tls_certificate_config_impl.h"
+
+namespace Envoy {
+namespace Secret {
+
+SdsApi::SdsApi(const LocalInfo::LocalInfo& local_info, Event::Dispatcher& dispatcher,
+               Runtime::RandomGenerator& random, Stats::Store& stats,
+               Upstream::ClusterManager& cluster_manager, Init::Manager& init_manager,
+               const envoy::api::v2::core::ConfigSource& sds_config, std::string sds_config_name,
+               std::function<void()> unregister_secret_provider)
+    : local_info_(local_info), dispatcher_(dispatcher), random_(random), stats_(stats),
+      cluster_manager_(cluster_manager), sds_config_(sds_config), sds_config_name_(sds_config_name),
+      secret_hash_(0), unregister_secret_provider_cb_(unregister_secret_provider) {
+  // TODO(JimmyCYJ): Implement chained_init_manager, so that multiple init_manager
+  // can be chained together to behave as one init_manager. In that way, we let
+  // two listeners which share same SdsApi to register at separate init managers, and
+  // each init manager has a chance to initialize its targets.
```

Comment:
I'm not sure I fully understand the implication here. If two listeners point to the same secret today, what happens? Does one listener just immediately initialize?

Reply:
In the current implementation,  yes.  The second listener will not wait, its initialization will finish right away since its init_manager is not used. Only the first one wait for the sds server response. 
The second listener will be ready prematurely, and start to listen on a port.   The connection will be reset before the SDS get the secret. 

I have a rough idea of how to solve this:  design a InitManagerChain; it can register multiple init_managers, but behaves like one init_manager.  Each indivicual init_manager's callbacks will be called when the combined_init_manager callback is called. 

That is what this TODO about.  like to implement it as a separate PR.

Reply:
Can you also file an issue? This seems a major issue, since you will not have warming on SDS work correctly otherwise.

Reply:
I have filed an issue here https://github.com/envoyproxy/envoy/issues/4224. Please take a look and let me know if any modification is required. Thanks.